ofc-tests-runner/run_ofc_tests.sh

```diff
@@ -0,0 +1,196 @@
+  fi
+}
+
+generate_actual() {
+  echo -e "-Generate actual"
+  local src_path="$1"
+  local output_path="$2"
+  local compiler="/opt/codethink-toolchain/bin/gfortran"
+  local compiler_args="-fdec -std=extra-legacy -frecursive -fno-automatic"
+  compile "$src_path" "$output_path" "$compiler" "$compiler_args"
+}
+
+generate_expected() {
+  echo -e "-Generate expected"
+  local src_path="$1"
+  local output_path="$2"
+
+  local src_name="$(basename $src_path)"
+  local src_dir="$(dirname $src_path)"
+  local stdout_path="$src_dir/stdout/$src_name"
+
+  ## Copy .expected file if exists, otherwise compile
+  if [ -f "$stdout_path" ]; then
+    echo -e "--Copy"
+    cp "$stdout_path" "$output_path"
+  else
+    local compiler="gfortran"
+    local compiler_args="-fdec -frecursive -fno-automatic"
+    compile "$src_path" "$output_path" "$compiler" "$compiler_args"
+  fi
+}
+
+compare_actual_with_expected() {
+  if $(diff "$1" "$2" &> "$3"); then
+    echo "pass"
+  else
+    echo "fail"
+  fi
+}
+
+run_tests() {
+  local source_paths="$1"
+  local output_dir="$2"
+  local actual_results_name="$3"
+
+  local output_generated_dir="$output_dir/generated"
+  local actual_results_path="$output_dir/$actual_results_name"
+
+  mkdir -p "$output_generated_dir"
+
+  local actual_results=''
+  for test_src_path in $source_paths; do
+    local test_src_abs_path="$(readlink -e $test_src_path)"
+    local test_name="$(basename $test_src_abs_path)"
+    local output_core_path="$output_generated_dir/$test_name"
+
+    echo -e "\n$test_name"
+    generate_expected "$test_src_abs_path" "$output_core_path.expected"
+    generate_actual "$test_src_abs_path" "$output_core_path.actual"
+    local test_result="$(compare_actual_with_expected "$output_core_path.actual" "$output_core_path.expected" "$output_core_path.diff")"
+    local test_line="$test_name:$test_result"
+    echo -e "-Result: $test_result"
+
+    actual_results="$actual_results$test_line\n"
+  done
+  echo -e "$actual_results" > "$actual_results_path"
+}
+
+compare_expected_with_actual_results() {
+  if ! stderr="$(diff $1 $2)"; then
+    echo -e "\nOFC tests failed"
+    echo "$stderr"
+    exit 1
+  fi
+}
+
+run_ofc_tests() {
+  OFC_TESTS_PROGRAMS_DIR="$OFC_TESTS_DIR/programs"
+  OFC_TESTS_NIST_DIR="$OFC_TESTS_PROGRAMS_DIR/nist"
+
+  OFC_OUTPUT_PROGRAMS_DIR="$OFC_OUTPUT_DIR/programs"
+  OFC_OUTPUT_NIST_DIR="$OFC_OUTPUT_DIR/nist"
+
+  ACTUAL_RESULTS_PROGRAMS_NAME="programs_actual_results"
+  ACTUAL_RESULTS_PROGRAMS_PATH="$OFC_OUTPUT_PROGRAMS_DIR/$ACTUAL_RESULTS_PROGRAMS_NAME"
+  ACTUAL_RESULTS_NIST_NAME="nist_actual_results"
+  ACTUAL_RESULTS_NIST_PATH="$OFC_OUTPUT_NIST_DIR/$ACTUAL_RESULTS_NIST_NAME"
+
+  SCRIPT_DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
+  EXPECTED_RESULTS_DIR="$SCRIPT_DIR/expected_test_results"
+  EXPECTED_RESULTS_PROGRAMS_PATH="$EXPECTED_RESULTS_DIR/programs_expected_results"
+  EXPECTED_RESULTS_NIST_PATH="$EXPECTED_RESULTS_DIR/nist_expected_results"
+
```

Comment:
I prefer to write the names of the variables which are part of the script and not in ENV in lower case and if they are not going to be out of this function declare them local, like in the other functions.